sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -3406,42 +3406,990 @@ lsc_format_ret(__ESIMD_NS::simd<T1, N> Vals) {
+///               simd<T, N> src0,
+///               simd_mask<1> pred = 1);                      // (lacc-au2-5)
+///
+/// simd<T, N>
+/// atomic_update(local_accessor lacc,
+///               simd_view<simd<uint32_t, N>, OffsetRegionTy> byte_offset,
+///               simd<T, N> src0,
+///               simd_view<simd<T, N>, RegionTy> src1,
+///               simd_mask<1> pred = 1);                      // (lacc-au2-6)
+///
+/// simd<T, N>
+/// atomic_update(local_accessor lacc,
+///               simd_view<simd<uint32_t, N>, OffsetRegionTy> byte_offset,
+///               simd_view<simd<T, N>, RegionTy> src0,
+///               simd<T, N> src1,
+///               simd_mask<1> pred = 1);                      // (lacc-au2-7)
+///
+/// simd<T, N>
+/// atomic_update(local_accessor lacc,
+///               simd_view<simd<uint32_t, N>, OffsetRegionTy> byte_offset,
+///               simd_view<simd<T, N>, RegionTy> src0,
+///               simd_view<simd<T, N>, RegionTy> src1,
+///               simd_mask<1> pred = 1);                      // (lacc-au2-8)
+///
+
+/// simd<T, N>
+/// slm_atomic_update(simd<uint32_t, N> byte_offset,
+///                   simd<T, N> src0, simd<T, N> src1,
+///                   simd_mask<N> mask = 1);                   /// (slm-au2-1)
+/// Atomically updates \c N memory locations in SLM indicated by
+/// a vector of offsets, and returns a vector of old
+/// values found at the memory locations before update.
+/// @tparam Op The atomic operation.
+/// @tparam T The vector element type.
+/// @tparam N The number of memory locations to update.
+/// @param byte_offset The vector of 32-bit offsets.
+/// @param src0 is the first atomic operand (new value).
+/// @param src1 is the second atomic operand (expected value).
+/// @param mask Operation mask, only locations with non-zero in the
+///   corresponding mask element are updated.
+/// @return A vector of the old values at the memory locations before the
+///   update.
+template <atomic_op Op, typename T, int N>
+__ESIMD_API std::enable_if_t<__ESIMD_DNS::get_num_args<Op>() == 2, simd<T, N>>
+slm_atomic_update(simd<uint32_t, N> byte_offset, simd<T, N> src0,
+                  simd<T, N> src1, simd_mask<N> mask = 1) {
+  // Only DG2/PVC support 8 bit data types.
+  if constexpr (sizeof(T) == 2 || sizeof(T) == 8 || !__ESIMD_DNS::isPowerOf2(N, 32)) {
+    // 2-argument lsc_atomic_update arguments order matches the standard one -
+    // expected value first, then new value. But atomic_update uses reverse
+    // order, hence the src1/src0 swap.
+    return detail::slm_atomic_update_impl<Op, T, N,
+                                          detail::lsc_data_size::default_size>(
+        byte_offset, src1, src0, mask);
+  } else {
+    detail::check_atomic<Op, T, N, 2>();
+    const auto si =
+        __ESIMD_NS::get_surface_index(detail::LocalAccessorMarker());
+    using Tx = detail::__raw_t<T>;
+    return __esimd_dword_atomic2<Op, Tx, N>(mask.data(), si, byte_offset.data(),
```

Comment:
Same comment as for 1-operand atomic_update.
Most likely we Tx!=T case never reaches this line. If so, `Tx` can be just removed. Otherwise, need to bit_cast src0 and src1 before call to __esimd_dword_atomic2().

Reply:
Done.